### ISSUE TITLE
Performance issues when starting the project

### ISSUE THREAD

**Issue by juanceborba** [mzheravin/exchange-core#53]
Hi Maksim, I hava a question about the core. We are using your core as a dependency. We are trying to run the project taking our first steps to know how it works. Our first issue is related to performance. When the project start our CPU consumption goes too high (more than 500% up). Is this normal? We are running the project in development mode, using a MACBook Pro Core I5 2.9Ghz. 

I´m using default configuration as your readme file says
        ExchangeConfiguration conf = ExchangeConfiguration.defaultBuilder().build();
If not, could be something related to the configuration?
Or could you suggest me where we can find the possible issue?
Thanks in advance, 
JC

**Comment by mzheravin** [comment]
Hi @juanceborba  
Yes, default configuration is expected to not load the system  in idle mode, because it uses "sleeping" instead of busy spins everywhere. However it consumes about 55% of CPU (just measured on my system), this is an old known issue, because some processors still use busy-spins.
However 500% looks like low latency configuration, where 5 threads are spinning, rather than basic performance configuration... maybe you could try to set it explicitly: 
```
 ExchangeConfiguration conf = ExchangeConfiguration.defaultBuilder().perfCfg(PerformanceConfiguration.baseBuilder().build()).build();
```
From my side, I will try to improve consumption for basic mode, so it will be close to 0%.

**Comment by mzheravin** [comment]
You can try recent version 0.5.0 - it shows very low consumption on my windows developer machine

**Comment by juanceborba** [comment]
Thanks Maksim for your quick reply. I already tried the recommended configuration and the problem still persists. I will try with the latest version right now!

**Comment by juanceborba** [comment]
Hi Maksim, I tried the new version and the performance is better, but the consumption of cpu is still very large. At the moment we are using two different infra environments.

1) Using docker. The consumption there is around 300%. We reduced the cpu use from 500% to 300% with your new version.

2) Without docker, the consumption in idle mode is around 70%. Consumption is 30% lower with the new version but still so large.

Either way, in both cases it is a lot of cpu consumption. Our live infrastructure will be assembled in Azure with Kubernetes and it would be very expensive to maintain it in this way. Also, I assume that when we actually start using it, consumption can go much higher. We will continue to see if we can make any improvements, I would really be interested in being able to use the core for our project.

I have another question for you. We are using the core as a dependency. Could be better to fork the project or is not a difference?

Thanks for your support! JC

**Comment by mzheravin** [comment]
As for performance I think it does not really matter if you run jar from maven central or compile yourself (unless the bytecode is different in the spinning part for some reason). But you could try. Anyway if you need to check recent code without waiting a release - you would have to fork/checkout.
I also ovserve high CPU usage under linux (no docker), as I mentioned. For Windows somehow it is very low though. 
There is one more option I want to try - Disruptor has blocker strategy https://github.com/LMAX-Exchange/disruptor/blob/master/src/main/java/com/lmax/disruptor/BlockingWaitStrategy.java - I will check and confirm if it works stable and does not cosume cpu.

**Comment by juanceborba** [comment]
Ok, understood. I will try to compile the project to analize it. Let me know if you have a progress with Disruptor. 

Another question, are you using Parity Trading engine as part of your project? (is another open source project that I found in the net) Or do you built the project from scratch? 

Thanks a lot!

**Comment by mzheravin** [comment]
Yes there is a dependency from parity trading indeed. It is used for reading NASDAQ market data  (ITCH format) for tests only. 
Exchange-core is not built from scratch - as you can see there are dependencies from industry leading projects. However, orders matching and risk control algrithms are authentic.

**Comment by mzheravin** [comment]
Hi @juanceborba could you please check latest release (0.5.1) CPU usage?

**Comment by mzheravin** [comment]
fixed with version 0.5.3
